src/sentry/static/sentry/app/views/eventsV2/saveQueryButton.tsx

```diff
@@ -0,0 +1,142 @@
+import React from 'react';
+import styled from 'react-emotion';
+import {browserHistory} from 'react-router';
+import {Location} from 'history';
+
+import {Client} from 'app/api';
+import {Organization} from 'app/types';
+import {t} from 'app/locale';
+import {
+  createSavedQuery,
+  updateSavedQuery,
+} from 'app/actionCreators/discoverSavedQueries';
+import {addSuccessMessage} from 'app/actionCreators/indicator';
+import DropdownControl from 'app/components/dropdownControl';
+import DropdownButton from 'app/components/dropdownButton';
+import Button from 'app/components/button';
+import Input from 'app/components/forms/input';
+import InlineSvg from 'app/components/inlineSvg';
+import space from 'app/styles/space';
+import withApi from 'app/utils/withApi';
+
+import EventView from './eventView';
+
+type Props = {
+  api: Client;
+  organization: Organization;
+  eventView: EventView;
+  location: Location;
+  isEditing: boolean;
+};
+
+type State = {
+  queryName: string;
+};
+
+class EventsSaveQueryButton extends React.Component<Props, State> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      queryName: props.isEditing ? props.eventView.name : '',
+    };
+  }
+
+  swallowEvent = (event: React.MouseEvent) => {
+    // Stop propagation for the input and container so
+    // people can interact with the inputs in the dropdown.
+    const capturedElements = ['LI', 'INPUT'];
+    if (
+      event.target instanceof Element &&
+      capturedElements.includes(event.target.nodeName)
+    ) {
+      event.preventDefault();
+      event.stopPropagation();
+    }
+  };
+
+  handleSave = () => {
+    const {api, eventView, organization, location, isEditing} = this.props;
+
+    const payload = eventView.toNewQuery();
+    if (this.state.queryName) {
+      payload.name = this.state.queryName;
+    }
+    let promise;
+    if (isEditing) {
+      promise = updateSavedQuery(api, organization.slug, payload);
+    } else {
+      promise = createSavedQuery(api, organization.slug, payload);
+    }
+    promise.then(saved => {
+      const view = EventView.fromSavedQuery(saved);
+      addSuccessMessage(t('Query saved'));
+
+      browserHistory.push({
```

Comment:
So browser history goes from edit -> updated query view, and we want to allow the user to hit back button to go back to edit mode?

Reply:
We don't want edit to be in the browser history. However, this is a temporary solution until Danny's work on the editable grid is done. The editable grid will likely require the edit state to be moved into a React.Context or reflux.